### PR TITLE
feat: YDoc migration script

### PR DESCRIPTION
**Pull request by thorseraq**
upstream issues:  #3732 

## Scope

From Editor Version: 0.0.0-20230420210727-85b7de79-nightly to latest

## TODO
- [x] Migrate from YDoc generated by previous AFFiNE Cloud to AFFiNE Client 0.8.0
- [x] Verify the migation result with BlockSuite playground
- [x] Insert migrated YDoc info into incoming AFFiNE Cloud storage

## Known issue

- Lines drawn by Brush(Pen) will lose
- Sometimes connector will lose, ie, when in old AFFiNE Cloud editor, connector source and target were not anchored with other elements

### Edgeless
image
image

### Blog
image

### CODE REVIEW

File: packages/doc-migration/package.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "doc-migration",
+  "packageManager": "yarn@3.6.0",
```
Comment (himself65) [inline packages/doc-migration/package.json]: no this

File: .yarn/patches/@blocksuite-store-npm-0.0.0-20230814155455-ceb5d5d8-nightly-8b4457adbc.patch [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+-        currentVersion,
+-        `previous version for flavour ${flavour} not found`
+-      );
++      if (!currentVersion) {
++        console.warn(`previous version for flavour ${flavour} not found`);
++        return;
++      }
+       this.upgradeBlock(flavour, currentVersion, block);
+     });
+   };
```
Comment (himself65) [inline .yarn/patches/@blocksuite-store-npm-0.0.0-20230814155455-ceb5d5d8-nightly-8b4457adbc.patch]: Please contribute to the upstream directly

**Review by himself65 — CHANGES_REQUESTED** [review]

**Review by himself65 — COMMENTED** [review]
I wonder if we can abstract migration logic to the top level so desktop, web, server can use the same logic

File: .yarn/patches/@blocksuite-store-npm-0.0.0-20230814155455-ceb5d5d8-nightly-8b4457adbc.patch (hunk 2 above)
Comment (thorseraq) [reply .yarn/patches/@blocksuite-store-npm-0.0.0-20230814155455-ceb5d5d8-nightly-8b4457adbc.patch]: Concerning when this migration part is running in the browser, throw might be a good behavior, I removed the patch and add a try catch in migration script

**Comment by thorseraq** [comment]
> I wonder if we can abstract migration logic to the top level so desktop, web, server can use the same logic

Currently the script is not well-abstracted enough, but I think it's doable, a quick idea looks:

- Most of migrations can be treated as normal CRDT operations, which is transparent to different platforms. For this part, we need a stable YDoc Meta structure(which we already have) to confirm the block version, and utils to migrate from version A -> version B, etc

- For the doc -> subdoc part, different storage provider in platforms need implement a common defined interface

**Pull request closed without merging**